test/cloudtest/test_secrets.py

```diff
@@ -56,3 +58,29 @@ def test_secrets(mz: MaterializeApplication) -> None:
     mz.environmentd.sql("DROP SECRET username CASCADE")
 
     wait(condition="delete", resource=f"secret/{secret}")
+
+
+# Tests that secrets deleted from the catalog but not from k8s are cleaned up on
+# envd startup.
+def test_orphaned_secrets(mz: MaterializeApplication) -> None:
+    mz.environmentd.sql("SET failpoints = 'drop_secrets=panic'")
+    mz.environmentd.sql("CREATE SECRET orphan AS '123'")
+
+    id = mz.environmentd.sql_query("SELECT id FROM mz_secrets WHERE name = 'orphan'")[
+        0
+    ][0]
+    assert id is not None
+    secret = f"user-managed-{id}"
+
+    # The failpoint should cause this to fail.
+    try:
+        mz.environmentd.sql("DROP SECRET orphan")
+        raise Exception("Unexpected success")
+    except InterfaceError:
+        pass
+
+    describe = mz.kubectl("describe", "secret", secret)
```

Comment:
Is this racy? Isn't `environmentd` restarting in the background and possibly deleting this secret at any point?

Reply:
Yes! Is there a way to disable the restart attempt?

Reply:
Nevermind: added another failpoint to prevent orphan cleanup before we want it.